CI: Support PR Preview

Shaikh-Ubaid opened the pull request:
This `PR` adds support of previewing a `pull request`. The built site is currently pushed into `gh-pages` branch at https://github.com/Shaikh-Ubaid/pull_request_preview.

Example PR: https://github.com/Shaikh-Ubaid/lcompilers_frontend/pull/2

Shaikh-Ubaid commented:
It seems the `Create comment about deployment` step failed. I guess the GitHub actions bot is unable to comment as the `CI` is not yet a part of master or because the `CI` might not be running in its own `GitHub remote`. Since, the comment works at https://github.com/Shaikh-Ubaid/lcompilers_frontend/pull/2, I am hoping it would work at  https://github.com/lfortran/lcompilers_frontend as well (I guess the pull request might need to be created at/from `origin/feature_branch`).

Shaikh-Ubaid commented:
This is ready. Please possibly review and please share feedback.

certik commented:
I would reuse the workflow that we have: the main purpose of the preview build is to ensure that it will work in production. For that we need to use the exact same steps, including checking out any production repositories etc. Only at the very end, instead of pushing into production, we push into the test repository.

Shaikh-Ubaid commented:
This is ready. Please possibly review and please share feedback.

Shaikh-Ubaid commented:
I transferred the `pull_request_preview` repository here https://github.com/lfortran/pull_request_preview

certik commented on ci/upload_site.sh:
```diff
@@ -0,0 +1,48 @@
+
+git_ref=${GITHUB_REF}
+
+if [[ ${git_ref} == "refs/heads/main" ]]; then
+    # Production version - pipeline triggered from main branch
+    deploy_repo_pull="https://github.com/lfortran/lcompilers_frontend.git"
+    deploy_repo_push="git@github.com:lfortran/lcompilers_frontend.git"
+else
+    # Test version - pipeline triggered from pull request
+    deploy_repo_pull="https://github.com/Shaikh-Ubaid/pull_request_preview.git"
+    deploy_repo_push="git@github.com:Shaikh-Ubaid/pull_request_preview.git"
+fi
```
Why not:

```suggestion
deploy_repo_pull="https://github.com/lfortran/lcompilers_frontend.git"
if [[ ${git_ref} == "refs/heads/main" ]]; then
    # Production version - pipeline triggered from main branch
    deploy_repo_push="git@github.com:lfortran/lcompilers_frontend.git"
else
    # Test version - pipeline triggered from pull request
    deploy_repo_push="git@github.com:Shaikh-Ubaid/pull_request_preview.git"
fi
```

Shaikh-Ubaid replied:
doubt: Please, could you possibly share if the above approach could lead to errors such as: 

(Not an exact error, just an example)
```
Failed to push refs: The repo is x commits ahead, use git pull
```

Shaikh-Ubaid replied:
If we pull the `pull_request_preview` and then update/push to it, I guess, we are (hopefully) avoiding or reducing the chances of `Failed to push refs` error.

certik replied:
Good point. I think in the past I did force push, but force pushing to a production repository might not be a good practice, since it can hide issues.

How are different PRs handled, or do all PRs push into just master, so they overwrite each other?

Shaikh-Ubaid replied:
>How are different PRs handled, or do all PRs push into just master, so they overwrite each other?

Yes, it seems they would overwrite each other.

Shaikh-Ubaid commented:
I updated the links to point to the new repository at https://github.com/lfortran/pull_request_preview

Shaikh-Ubaid commented:
Ondrej Sir, please, could you possibly share if it would be possible for you to setup the needed `SSH` Keys, please?

Shaikh-Ubaid commented:
Also, I updated the preview link here https://github.com/lfortran/lcompilers_frontend/pull/38/commits/893812ea922a06d07022cc041fe12c5055153fea to https://preview.dev.lfortran.org. Please, could someone possibly share if the link is fine?

Shaikh-Ubaid commented:
>Also, I updated the preview link here https://github.com/lfortran/lcompilers_frontend/commit/893812ea922a06d07022cc041fe12c5055153fea to https://preview.dev.lfortran.org/. Please, could someone possibly share if the link is fine and if it is possible to use it?

If we use a link which has some path after domain (for example: https://dev.lfortran.org/pull_request_preview), we would be needing to update the `basepath` as in here https://github.com/lfortran/lcompilers_frontend/pull/38/commits/0e80d775b6edc28d68355cec526c6ea60cbdcc7e

Shaikh-Ubaid commented:
This is ready. Please possibly review and please share feedback.

certik commented on .github/workflows/deploy.yml:
```diff
@@ -43,28 +46,17 @@ jobs:
+      with:
+        issue-number: ${{ github.event.pull_request.number }}
+        body: |
+          Your site is deployed at https://preview.dev.lfortran.org/
```
I don't know how subdomains work --- we can also just use regular github pages, to make it clear it is just a PR preview.

Shaikh-Ubaid replied:
>I don't know how subdomains work --- we can also just use regular github pages, to make it clear it is just a PR preview.

It seems when we use the regular github pages, the site would be available at https://lfortran.github.io/pull_request_preview and thus adding the path `/pull_request_preview` to the url. Supporting the path might be challenging as I shared here https://github.com/lfortran/lcompilers_frontend/pull/38#issuecomment-1222102903

certik commented on ci/upload_site.sh:
```diff
@@ -0,0 +1,48 @@
+    echo "Note: GIT_PR_PREVIEW_PRIVATE_SSH_KEY is empty, skipping..."
+    exit 0
+fi
+ssh-add <(echo "$GIT_PR_PREVIEW_PRIVATE_SSH_KEY" | base64 -d)
```
We already have ssh keys setup with a different variable (I believe), so this PR will only use `GIT_PR_PREVIEW_PRIVATE_SSH_KEY` to push even to production?

I think we should use the production keys for production and preview keys for previews, what do you think?

Shaikh-Ubaid replied:
>We already have ssh keys setup with a different variable (I believe), so this PR will only use GIT_PR_PREVIEW_PRIVATE_SSH_KEY to push even to production?

It seems there are no `ssh` keys required to push to production (as we discussed here https://github.com/lfortran/lcompilers_frontend/pull/1#discussion_r945094525 and also tried/tested it by removing the `SSH` key setup step here https://github.com/lfortran/lcompilers_frontend/commit/c6cee87fe0a64de5aa9ca7495fe2c1fc39dfec8b). We currently need the `SSH` keys for pushing the test version to `pull_request_preview` (as `pull_request_preview` is a separate repository). During production deployment, even if the `SSH` keys would be available, I guess/think that they would (hopefully) not be used.

Shaikh-Ubaid replied:
>We already have ssh keys setup with a different variable (I believe)

Please, could you possibly share if it would be possible to rename it and setup its `public` key at https://github.com/lfortran/pull_request_preview, please

certik replied:
I see. It's a little weird that it would quit to push to production if GIT_PR_PREVIEW_PRIVATE_SSH_KEY is empty. But I guess that's fine. Would it make sense to put this into `if [[ ${git_ref} != "refs/heads/main" ]]; then` if statement?

Shaikh-Ubaid replied:
Yes, got it.

Shaikh-Ubaid commented:
>The way to do it is to use a prefix directory with the PR number.

Got it. Thank you so much for the guidance.

Shaikh-Ubaid commented:
>I came across https://github.com/nf-core/tools/issues/765 and https://securitylab.github.com/research/github-actions-preventing-pwn-requests/. It seems to me that we can support previewing pull requests even when the PRs are submitted from a fork.

I would like to try this out. Please, could you possibly share if we shall create three separate workflows as follows:
- first workflow runs only on `push` to `main` branch event
    - It is the same one that we currently have in the master
- second workflow runs only on `pull request` event
    - It runs in an unprivileged environment and builds the project and uploads the built files 
- third workflow runs only on after completion of the second workflow
    - It runs in a privileged environment
    - It downloads the built files of the second workflow and pushes them to the `pull_request_preview` repository 
    - It comments on the pull request with the deployment link

certik commented on ci/upload_site.sh:
```diff
@@ -0,0 +1,50 @@
+        echo "Note: GIT_PR_PREVIEW_PRIVATE_SSH_KEY is empty, skipping..."
+        exit 0
+    fi
+fi
+ssh-add <(echo "$GIT_PR_PREVIEW_PRIVATE_SSH_KEY" | base64 -d)
```
I think this needs to be:
```suggestion
    ssh-add <(echo "$GIT_PR_PREVIEW_PRIVATE_SSH_KEY" | base64 -d)
fi
```

certik reviewed (approved):
I think this is fine. When you merge, carefully watch the builds, and test the final site. If something goes wrong, then first revert the deploy commit in the deploy repo to fix the site to the previous state, then try to fix it with new PRs.

Shaikh-Ubaid commented:
>The way to do it is to use a prefix directory with the PR number.

Ondrej Sir, it seems,  the workflow currently overwrites the `gh-pages` branch for every change in different pull-requests.

Shaikh-Ubaid commented:
>I think this is fine. When you merge, carefully watch the builds, and test the final site. If something goes wrong, then first revert the deploy commit in the deploy repo to fix the site to the previous state, then try to fix it with new PRs.

Got it. 

I looked into the changes of this pull request again and they seem fine to me. Since, this seems to be approved, I am merging this PR for now.

Shaikh-Ubaid commented:
The CI failed to push to `gh-pages` branch of https://github.com/lfortran/lcompilers_frontend. It seems that since we used the `https` url to clone the repository https://github.com/lfortran/lcompilers_frontend, it is needing ssh keys to push to the same repository.

I am submitting a new PR which would use the ssh url for the same repository https://github.com/lfortran/lcompilers_frontend for the commit in the main branch.

Since, the push to `gh-pages` branch of https://github.com/lfortran/lcompilers_frontend failed, the site should be hopefully intact. It also seems to be.

Shaikh-Ubaid commented:
>I am submitting a new PR which would use the ssh url for the same repository https://github.com/lfortran/lcompilers_frontend for the commit in the main branch.

I fixed the `CI` failure issue while pushing to `gh-pages` branch of https://github.com/lfortran/lcompilers_frontend. 

Also, while updating the `latest_commit`, I fixed the issues discovered related to this `pull_request_preview` feature in https://github.com/lfortran/lcompilers_frontend/pull/44.

certik commented:
Thanks for fixing it!
